### PR TITLE
[BUG] Use models literal StructuredDataset to enable sd bypass task

### PR DESCRIPTION
**Pull request by JiangJiaWei1103**
## Tracking issue
Closes [flyteorg/flyte#5956](https://github.com/flyteorg/flyte/issues/5956).

## Why are the changes needed?
For a flyte task with python `StructuredDataset` as an input from `dataclass` attribute access (i.e., `dc.sd`), it fails to convert the task output `LiteralMap` back to flyte idl. On the highest level, the process of calling `_dispatch_execute` ([here](https://github.com/flyteorg/flytekit/blob/master/flytekit/bin/entrypoint.py#L134)) is illustrated as follows:

pythontask_sd_io

As can be observed, the problem (marked as orange) occurs when the input `LiteralMap` is converted to python natives, during which a `Literal` is built with python native `StructuredDataset`, instead of using the correct `literals.StructuredDataset` (please refer to this [code snippet](https://github.com/flyteorg/flytekit/blob/2e40e76d8a5614a7eedcac8399ca877e64612640/flytekit/types/structured/structured_dataset.py#L800-L809)).

This makes conversion from the task output `LiteralMap` back to flyte idl fail because python native `StructuredDataset` has no `to_flyte_idl` method.

## What changes were proposed in this pull request?
Directly construct a `literals.StructuredDataset` within `dict_to_structured_dataset` method of `StructuredDatasetTransformerEngine`.

## How was this patch tested?
1. Execute the command run in the K8s pod, as suggested by @pingsutw:
    * Describe the pod executing the remote task in the [reported issue](https://github.com/flyteorg/flyte/issues/5956).
    * Copy the command in the pod and reproduce locally.

After fixing, `outputs.pb` now can be successfully written to s3 storage.
Screenshot 2024-11-26 at 9 52 57 PM

Screenshot 2024-11-26 at 9 54 19 PM

2. Do remote sandbox test with `flytekit` installed by commit sha in this PR.
```python
from dataclasses import dataclass, field

from flytekit import task, workflow, ImageSpec
from flytekit.types.structured import StructuredDataset


# Build docker image and push to the local registry
flytekit_hash = "d5d8620cc0494a148adb6bb4a13978d810d9673a"  # Refer to this PR
flytekit = f"git+https://github.com/flyteorg/flytekit.git@{flytekit_hash}"
image = ImageSpec(
    packages=[
        flytekit,
        "pandas",
        "pyarrow"
    ],
    apt_packages=["git"],
    registry="localhost:30000",
)


@dataclass
class DC:
    sd: StructuredDataset = field(
        default_factory=lambda: StructuredDataset(
            uri="s3://my-s3-bucket/s3_flyte_dir/df.parquet", 
            file_format="parquet"
        )
    )


@task(container_image=image)
def t_sd_attr(sd: StructuredDataset) -> StructuredDataset:
    return sd


@workflow
def wf(dc: DC):
    t_sd_attr(sd=dc.sd)
```

Run the following command:
```
pyflyte run --remote reprod.py wf --dc '{"dc": {"sd": {"uri": "s3://my-s3-bucket/s3_flyte_dir/df.parquet", "file_format": "parquet"}}}'
```

The remote sandbox test result is shown as follows:

Screenshot 2024-11-26 at 10 14 03 PM

### Setup process
```
git clone https://github.com/flyteorg/flytekit.git
gh pr checkout 2954
make setup && pip install -e .
```

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/2914

### CODE REVIEW

File: flytekit/types/structured/structured_dataset.py [hunk 1]
```diff
@@ -197,12 +197,25 @@ def return_sd() -> StructuredDataset:
             return df
 
         For details, please refer to this issue: https://github.com/flyteorg/flyte/issues/5954.
+
+        2. Need access to self._literal_sd when converting task output back to flyteidl, please see:
+        https://github.com/flyteorg/flytekit/blob/master/flytekit/bin/entrypoint.py#L326
```
Comment (Future-Outlier) [inline flytekit/types/structured/structured_dataset.py:202]: Can we use github permanent link?

Comment (JiangJiaWei1103) [reply flytekit/types/structured/structured_dataset.py:202]: Sure, let me fix it. Thanks!

**Comment by JiangJiaWei1103** [comment]
### Follow-up

The last commit aims at simplify the building logic of `sd_literal`. Instead of constructing a dummy python native `StructuredDataset` then calling `set_literal`, we directly build `sd_literal` here. Thanks for @Future-Outlier 's suggestion.

The next steps are:
1. Make sure this patch passes the sandbox test using the image built with the last commit sha.
2. Handle the empty-str `file_format` issue.

**Comment by Future-Outlier** [comment]
> ### Follow-up
> The last commit aims at simplifying the building logic of `sd_literal`. Instead of constructing a dummy python native `StructuredDataset` then calling `set_literal`, we directly build `sd_literal` here. Thanks for @Future-Outlier 's suggestion.
> 
> The next steps are:
> 
> 1. Make sure this patch passes the sandbox test using the image built with the last commit sha.
> 2. Handle the empty-string `file_format` issue.

1. you can use single binary to test it, this will be quicker.
2. this will help a lot, thank you man

**Comment by JiangJiaWei1103** [comment]
Hi @Future-Outlier,

The last commit passes the test example you provided me yesterday, as shown below:

Screenshot 2024-11-29 at 11 08 33 PM

For the first question, I do use a **single binary** to run remote tests as suggested in the [official setup](https://docs.flyte.org/en/latest/community/contribute/contribute_code.html#how-to-setup-dev-environment-for-flyteidl-flyteadmin-flyteplugins-flytepropeller-datacatalog-and-flytestdlib). In step 2, `flytectl demo start --dev` is used to start the k3s cluster and this environment is referred to as **Flyte sandbox**, so I just call it a **sandbox test**. Sorry for the dumb question here, but I would like to further clarify the difference and refine my understanding. Thanks very much!

For the empty-string `file_format`, I suspect this information is erased before `inputs.pb` is loaded from s3. Following shows the loaded input proto:
Screenshot 2024-11-29 at 11 27 50 PM

As can be observed, the information of `file_format` is absent. Hence, I think I'll continue tracing the serialization and registration process to find out what's going on.

**Comment by JiangJiaWei1103** [comment]
### Further Discussion - Logic behind `extract_cols_and_format`

For the last commit, we drop the redundant handling of column information of `StructuredDataset` in `dict_to_structured_dataset`. First of all, `extract_cols_and_format` isn't exclusive to deprecated `FlyteSchema`. It's used to iterate through `Annotation` and extract the information like base type, column information, etc. Following demonstrates a simple example about extracting one column out of a `pd.DataFrame` and returning one-column df:

```python
all_cols = kwtypes(name=str, height=int)
col = kwtypes(height=int)


@task(container_image=image)
def t1(sd: Annotated[StructuredDataset, all_cols]) -> Annotated[StructuredDataset, col]:
    df = sd.open(pd.DataFrame).all()

    return StructuredDataset(dataframe=df)
```

For these cases, `async_to_python_value`  can already handle column information, as can be seen [here](https://github.com/flyteorg/flytekit/blob/92d87a88a0a7a9969a3237a83a7008a6768748a2/flytekit/types/structured/structured_dataset.py#L916-L917) and [here](https://github.com/flyteorg/flytekit/blob/92d87a88a0a7a9969a3237a83a7008a6768748a2/flytekit/types/structured/structured_dataset.py#L958-L970). Hence, we just drop the redundant handling.

**Comment by JiangJiaWei1103** [comment]
Hi @Future-Outlier,

For the empty-string `file_format` problem, I plan to open another issue which will contain thorough experiments and observations this week!

**Review by Future-Outlier — APPROVED** [review]
Really Nice PR.
I can tell you put lots of efforts.
Let's make Flyte GREAT!

**Comment by JiangJiaWei1103** [comment]
Thanks for your support and patience, bro!

Let's move on and make Flyte better!!

**Pull request merged**